## Опиcание цели исследования

Коронавирус застал мир врасплох, изменив привычный порядок вещей. В свободное время жители городов больше не выходят на улицу, не посещают кафе и торговые центры. Зато стало больше времени для книг. Это заметили стартаперы — и бросились создавать приложения для тех, кто любит читать.

Ваша компания решила быть на волне и купила крупный сервис для чтения книг по подписке. Ваша первая задача как аналитика — проанализировать базу данных. В ней — информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Эти данные помогут сформулировать ценностное предложение для нового продукта.

## Задачи исследования

Для того, чтобы проанализировать базу данных вы полним следующие задачи:

- Посчитать, сколько книг вышло после 1 января 2000 года;
- Для каждой книги посчитайть количество обзоров и среднюю оценку;
- Определить издательство, которое выпустило наибольшее число книг толще 50 страниц — так мы исключим из анализа брошюры;
- Определить автора с самой высокой средней оценкой книг — учитывать только книги с 50 и более оценками;
- Посчитайть среднее количество обзоров от пользователей, которые поставили больше 48 оценок.


## Описание данных

Таблица books

Содержит данные о книгах:

* book_id — идентификатор книги;
* author_id — идентификатор автора;
* title — название книги;
* num_pages — количество страниц;
* publication_date — дата публикации книги;
* publisher_id — идентификатор издателя.

Таблица authors

Содержит данные об авторах:

* author_id — идентификатор автора;
* author — имя автора.

Таблица publishers

Содержит данные об издательствах:

* publisher_id — идентификатор издательства;
* publisher — название издательства;

Таблица ratings

Содержит данные о пользовательских оценках книг:

* rating_id — идентификатор оценки;
* book_id — идентификатор книги;
* username — имя пользователя, оставившего оценку;
* rating — оценка книги.

Таблица reviews

Содержит данные о пользовательских обзорах на книги:

* review_id — идентификатор обзора;
* book_id — идентификатор книги;
* username — имя пользователя, написавшего обзор;
* text — текст обзора.

## Структура базы данных:

In [1]:
# импортируем библиотеки
import pandas as pd
from sqlalchemy import text, create_engine
import datetime as dt


In [2]:
# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
'pwd': 'Sdf4$2;d-d30pp', # пароль
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # порт подключения
'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{user}:{pwd}@{host}:{port}/{db}'.format(**db_config)
# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'})


In [3]:
# создадим функцию для запросов к данным
def select(sql):
    return pd.io.sql.read_sql(sql, con = engine)

In [4]:
# получаем сырые данные
query = '''
            SELECT * FROM books
        '''

In [5]:
books = pd.io.sql.read_sql('SELECT * FROM books', con = engine)
books

book_id  author_id                                              title  \
0          1        546                                       'Salem's Lot   
1          2        465                 1 000 Places to See Before You Die   
2          3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3          4         82  1491: New Revelations of the Americas Before C...   
4          5        125                                               1776   
..       ...        ...                                                ...   
995      996        571           Wyrd Sisters (Discworld  #6; Witches #2)   
996      997        454                        Xenocide (Ender's Saga  #3)   
997      998        201                                    Year of Wonders   
998      999         94                        You Suck (A Love Story  #2)   
999     1000        509  Zen and the Art of Motorcycle Maintenance: An ...   

     num_pages publication_date  publisher_id  
0          594       2005-11-01            93  
1          992       2003-05-22           336  
2          322       2010-12-21           135  
3          541       2006-10-10           309  
4          386       2006-07-04           268  
..         ...              ...           ...  
995        265       2001-02-06           147  
996        592       1996-07-15           297  
997        358       2002-04-30           212  
998        328       2007-01-16           331  
999        540       2006-04-25           143  

[1000 rows x 6 columns]

In [6]:
# получаем сырые данные
query = '''
            SELECT * FROM authors
        '''

In [7]:
authors = pd.io.sql.read_sql('SELECT * FROM authors', con = engine)
authors

author_id                                author
0            1                            A.S. Byatt
1            2        Aesop/Laura Harris/Laura Gibbs
2            3                       Agatha Christie
3            4                         Alan Brennert
4            5              Alan Moore/David   Lloyd
..         ...                                   ...
631        632         William Strunk Jr./E.B. White
632        633                           Zadie Smith
633        634                 Zilpha Keatley Snyder
634        635                    Zora Neale Hurston
635        636  Åsne Seierstad/Ingrid Christopherson

[636 rows x 2 columns]

In [8]:
# получаем сырые данные
query = '''
            SELECT * FROM publishers
        '''

In [9]:
publishers = pd.io.sql.read_sql('SELECT * FROM publishers', con = engine)
publishers

publisher_id                          publisher
0               1                                Ace
1               2                           Ace Book
2               3                          Ace Books
3               4                      Ace Hardcover
4               5  Addison Wesley Publishing Company
..            ...                                ...
335           336         Workman Publishing Company
336           337                         Wyatt Book
337           338              Yale University Press
338           339                           Yearling
339           340                     Yearling Books

[340 rows x 2 columns]

In [10]:
# получаем сырые данные
query = '''
            SELECT * FROM ratings
        '''

In [11]:
ratings = pd.io.sql.read_sql('SELECT * FROM ratings', con = engine)
ratings

rating_id  book_id        username  rating
0             1        1      ryanfranco       4
1             2        1   grantpatricia       2
2             3        1    brandtandrea       5
3             4        2        lorichen       3
4             5        2     mariokeller       2
...         ...      ...             ...     ...
6451       6452     1000  carolrodriguez       4
6452       6453     1000         wendy18       4
6453       6454     1000      jarvispaul       5
6454       6455     1000           zross       2
6455       6456     1000         fharris       5

[6456 rows x 4 columns]

In [12]:
# получаем сырые данные
query = '''
            SELECT * FROM reviews
        '''

In [13]:
reviews = pd.io.sql.read_sql('SELECT * FROM reviews', con = engine)
reviews

review_id  book_id        username  \
0             1        1    brandtandrea   
1             2        1      ryanfranco   
2             3        2        lorichen   
3             4        3   johnsonamanda   
4             5        3     scotttamara   
...         ...      ...             ...   
2788       2789      999      martinadam   
2789       2790     1000         wknight   
2790       2791     1000  carolrodriguez   
2791       2792     1000         wendy18   
2792       2793     1000      jarvispaul   

                                                   text  
0     Mention society tell send professor analysis. ...  
1     Foot glass pretty audience hit themselves. Amo...  
2     Listen treat keep worry. Miss husband tax but ...  
3     Finally month interesting blue could nature cu...  
4     Nation purpose heavy give wait song will. List...  
...                                                 ...  
2788  Later hospital turn easy community. Fact same ...  
2789   Change lose answer close pressure. Spend so now.  
2790  Authority go who television entire hair guy po...  
2791  Or western offer wonder ask. More hear phone f...  
2792  Republican staff bit eat material measure plan...  

[2793 rows x 4 columns]

In [14]:
# изучим данные
list_df = (books, authors, publishers, ratings, reviews)
for list_df in list_df:
    print('\033[31m' + '______________info_____________________' + '\033[0m')
    display(list_df.info())
    print('\033[31m' + '______________describe_________________' + '\033[0m')
    display(list_df.describe().T)
    print('\033[31m' + '_____________isna______________________' + '\033[0m')
    display(list_df.isna().sum().reset_index())
    print('\033[31m' + '_______Поиск дубликатов________________' + '\033[0m')
    display(list_df.duplicated().sum())
    print('\033[31m' + '__количество пропущенных значений по каждому из столбцов__' + '\033[0m')
    display(list_df.isnull().mean().reset_index())
    print('\033[34m' + '_________________________конец изучения файла____________________________________' + '\033[0m')
    print('')

______________info_____________________
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 6 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   book_id           1000 non-null   int64 
 1   author_id         1000 non-null   int64 
 2   title             1000 non-null   object
 3   num_pages         1000 non-null   int64 
 4   publication_date  1000 non-null   object
 5   publisher_id      1000 non-null   int64 
dtypes: int64(4), object(2)
memory usage: 47.0+ KB


None

______________describe_________________


count     mean         std   min     25%    50%     75%     max
book_id       1000.0  500.500  288.819436   1.0  250.75  500.5  750.25  1000.0
author_id     1000.0  320.417  181.620172   1.0  162.75  316.5  481.00   636.0
num_pages     1000.0  389.111  229.390140  14.0  249.00  352.0  453.00  2690.0
publisher_id  1000.0  171.270   99.082685   1.0   83.00  177.5  258.00   340.0

_____________isna______________________


index  0
0           book_id  0
1         author_id  0
2             title  0
3         num_pages  0
4  publication_date  0
5      publisher_id  0

_______Поиск дубликатов________________


0

__количество пропущенных значений по каждому из столбцов__


index    0
0           book_id  0.0
1         author_id  0.0
2             title  0.0
3         num_pages  0.0
4  publication_date  0.0
5      publisher_id  0.0

_________________________конец изучения файла____________________________________

______________info_____________________
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 636 entries, 0 to 635
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   author_id  636 non-null    int64 
 1   author     636 non-null    object
dtypes: int64(1), object(1)
memory usage: 10.1+ KB


None

______________describe_________________


count   mean         std  min     25%    50%     75%    max
author_id  636.0  318.5  183.741666  1.0  159.75  318.5  477.25  636.0

_____________isna______________________


index  0
0  author_id  0
1     author  0

_______Поиск дубликатов________________


0

__количество пропущенных значений по каждому из столбцов__


index    0
0  author_id  0.0
1     author  0.0

_________________________конец изучения файла____________________________________

______________info_____________________
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 340 entries, 0 to 339
Data columns (total 2 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   publisher_id  340 non-null    int64 
 1   publisher     340 non-null    object
dtypes: int64(1), object(1)
memory usage: 5.4+ KB


None

______________describe_________________


count   mean        std  min    25%    50%     75%    max
publisher_id  340.0  170.5  98.293777  1.0  85.75  170.5  255.25  340.0

_____________isna______________________


index  0
0  publisher_id  0
1     publisher  0

_______Поиск дубликатов________________


0

__количество пропущенных значений по каждому из столбцов__


index    0
0  publisher_id  0.0
1     publisher  0.0

_________________________конец изучения файла____________________________________

______________info_____________________
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6456 entries, 0 to 6455
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   rating_id  6456 non-null   int64 
 1   book_id    6456 non-null   int64 
 2   username   6456 non-null   object
 3   rating     6456 non-null   int64 
dtypes: int64(3), object(1)
memory usage: 201.9+ KB


None

______________describe_________________


count         mean          std  min      25%     50%      75%  \
rating_id  6456.0  3228.500000  1863.831001  1.0  1614.75  3228.5  4842.25   
book_id    6456.0   510.574195   284.141636  1.0   291.00   506.0   750.00   
rating     6456.0     3.928284     0.943303  1.0     3.00     4.0     5.00   

              max  
rating_id  6456.0  
book_id    1000.0  
rating        5.0

_____________isna______________________


index  0
0  rating_id  0
1    book_id  0
2   username  0
3     rating  0

_______Поиск дубликатов________________


0

__количество пропущенных значений по каждому из столбцов__


index    0
0  rating_id  0.0
1    book_id  0.0
2   username  0.0
3     rating  0.0

_________________________конец изучения файла____________________________________

______________info_____________________
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2793 entries, 0 to 2792
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   review_id  2793 non-null   int64 
 1   book_id    2793 non-null   int64 
 2   username   2793 non-null   object
 3   text       2793 non-null   object
dtypes: int64(2), object(2)
memory usage: 87.4+ KB


None

______________describe_________________


count         mean         std  min    25%     50%     75%     max
review_id  2793.0  1397.000000  806.413976  1.0  699.0  1397.0  2095.0  2793.0
book_id    2793.0   504.693161  288.472931  1.0  259.0   505.0   753.0  1000.0

_____________isna______________________


index  0
0  review_id  0
1    book_id  0
2   username  0
3       text  0

_______Поиск дубликатов________________


0

__количество пропущенных значений по каждому из столбцов__


index    0
0  review_id  0.0
1    book_id  0.0
2   username  0.0
3       text  0.0

_________________________конец изучения файла____________________________________



#### Во всех таблицах разное количество строк, пропуски и дубликаты отсутствуют

Тип в колонке publication_date - object 

In [15]:
# изменим тип данных на дату
books['publication_date'] = pd.to_datetime(books['publication_date'])

In [16]:
# проверяем
books.dtypes

book_id                      int64
author_id                    int64
title                       object
num_pages                    int64
publication_date    datetime64[ns]
publisher_id                 int64
dtype: object

## Посчитаем, сколько книг вышло после 1 января 2000 года

In [17]:
query = '''
        SELECT 
        COUNT(book_id) as sum_books ----количество книг
        FROM books
        WHERE publication_date>'2000-01-01'
        '''
pd.read_sql(query, con = engine)

df = pd.io.sql.read_sql(query, con = engine)
df.columns = ['Количество книг']
df

Количество книг
0              819

#### Таких книг 819 ед.

## Для каждой книги посчитаем количество обзоров и среднюю оценку

In [18]:
query = '''
    SELECT b.book_id, b.title, 
    COUNT(DISTINCT r.review_id) as sum_rev, ----количество обзоров
    AVG(rt.rating) as avg_rating
    FROM BOOKS b 
        LEFT JOIN ratings rt ON rt.book_id=b.book_id 
        LEFT JOIN reviews r ON r.book_id=rt.book_id
    GROUP BY b.book_id
    ORDER BY sum_rev DESC
    '''

df = pd.io.sql.read_sql(query, con = engine)
df.columns = ['Идентификационный №','Название', 'Количество обзоров', 'Средний рейтинг']
df

Идентификационный №                                           Название  \
0                    948                            Twilight (Twilight  #1)   
1                    963                                Water for Elephants   
2                    734                                   The Glass Castle   
3                    302  Harry Potter and the Prisoner of Azkaban (Harr...   
4                    695  The Curious Incident of the Dog in the Night-Time   
..                   ...                                                ...   
995                   83    Anne Rice's The Vampire Lestat: A Graphic Novel   
996                  808                            The Natural Way to Draw   
997                  672   The Cat in the Hat and Other Dr. Seuss Favorites   
998                  221                          Essential Tales and Poems   
999                  191  Disney's Beauty and the Beast (A Little Golden...   

     Количество обзоров  Средний рейтинг  
0                     7         3.662500  
1                     6         3.977273  
2                     6         4.206897  
3                     6         4.414634  
4                     6         4.081081  
..                  ...              ...  
995                   0         3.666667  
996                   0         3.000000  
997                   0         5.000000  
998                   0         4.000000  
999                   0         4.000000  

[1000 rows x 4 columns]

#### Если брать за основу "количество обзоров", то в лидерах Twilight (Twilight #1) с 7 обзорами

## Определим издательство, которое выпустило наибольшее число книг 

In [19]:
query = '''
        SELECT p.publisher, ---- название издательства 
        COUNT (b.book_id) as sum_book --- количество книг 
        FROM books b 
            LEFT JOIN  publishers p ON b.publisher_id=p.publisher_id 
        WHERE b.num_pages>50   
        GROUP BY p.publisher
        ORDER BY sum_book DESC
        LIMIT 1
    '''
pd.read_sql(query, con = engine)

df = pd.io.sql.read_sql(query, con = engine)
df.columns = ['Издательство','Количество книг']
df

Издательство  Количество книг
0  Penguin Books               42

#### Таким издательством стало Penguin Books с 42 книгами (книга - издание, в котором больше 50 стр.)

## Определим автора с самой высокой средней оценкой книги

In [20]:
query  = '''
        SELECT a.author,             
        AVG(rt.rating) AS avg_rating ----cредний рейтинг
        FROM 
         books b
        LEFT JOIN authors a ON b.author_id = a.author_id
        LEFT JOIN ratings rt ON rt.book_id = b.book_id
        WHERE b.book_id IN (
            SELECT b.book_id FROM books b    
                LEFT JOIN ratings rt ON rt.book_id = b.book_id     
                     GROUP BY
                          b.book_id
                     HAVING 
                          COUNT(rt.rating_id) >= 50
                      )
                 GROUP BY
                    a.author  
                 ORDER BY 
                    avg_rating DESC
                 LIMIT 5    
       '''
pd.read_sql(query, con = engine)

df = pd.io.sql.read_sql(query, con = engine)
df.columns = ['Автор','Средний рейтинг']
df

Автор  Средний рейтинг
0         J.K. Rowling/Mary GrandPré         4.287097
1  Markus Zusak/Cao Xuân Việt Khương         4.264151
2                     J.R.R. Tolkien         4.246914
3                  Louisa May Alcott         4.192308
4                       Rick Riordan         4.080645

#### автор с самой высокой средней оценкой книг (50 и более оценок): J.K. Rowling

## Посчитаем среднее количество обзоров от пользователей

In [21]:
query = '''
     SELECT SUM(a.cnt) / COUNT(a.username) AS avg_cnt_reviews
     FROM
        (SELECT r.username,
            COUNT(r.review_id) AS cnt ---- количество оценок 
                FROM
                    reviews r 
                WHERE r.username IN  (SELECT rt.username FROM ratings rt
                     LEFT JOIN books b ON b.book_id = rt.book_id
                     GROUP BY
                         rt.username
                     HAVING
                         COUNT(rt.rating_id) > 48)
                GROUP BY r.username) AS a

       '''
pd.read_sql(query, con = engine).round()

df = pd.io.sql.read_sql(query, con = engine)
df.columns = ['Среднее количество обзоров']
df

Среднее количество обзоров
0                        24.0

#### Пользователи оставившие более 50 оценок, в среднем написали 24 обзора

## Вывод:

- С января 2000 г. вышло 819 книг
- Больше всего обзоров у книги Twilight(Twilight #1)  - 7 обзоров со средней оценкой 3,7
- Лидирующее издательство по выпуску книг - Penguin Books
- Самый рейтинговый писатель - J.K. Rowling.
- Среднее количество обзоров на пользователя с более чем 48 оценками - 24